### ISSUE TITLE
audioconvert standalone node doesn't work the first time it is linked

### ISSUE THREAD

**Issue by gkiagia** [PipeWire/pipewire#178]
Porting the design that was previously described in https://github.com/PipeWire/pipewire/pull/167 to a very recent pipewire version, we are now instantiating the spa audioconvert plugin through the spa node factory and then linking it to the adapter that contains the device node.

What we observe is that the very first time that this pipeline is linked, there is no audio. Apparently this is an issue of the audioconvert node that sits between the application and the adapter:

In `audioconvert.c: impl_node_port_set_param()` there is a call to `clean_convert(this);` that is triggered always if the PortConfig of the port is "convert". Inside `clean_convert`, the code clears the `this->buffers_set[...]` flags, which eventually results in buffers not being properly configured, because the order of the calls is the following:

(I have added these extra printouts for debugging)

```
[W][000075437.013805][audioconvert.c:957 impl_node_port_set_param()] 0x7f799b516038: SET PARAM:4 DIR:0 param:0x55d8a66d36c0
[W][000075437.013910][audioconvert.c:971 impl_node_port_set_param()] 0x7f799b516038:     clean 0, res 0
[W][000075437.013954][audioconvert.c:957 impl_node_port_set_param()] 0x7f799b516038: SET PARAM:4 DIR:0 param:0x55d8a66d37d0
[W][000075437.013981][audioconvert.c:971 impl_node_port_set_param()] 0x7f799b516038:     clean 0, res 0
[W][000075437.016070][audioconvert.c:1008 impl_node_port_use_buffers()] 0x7f799b516038: USE BUFFERS 0x55d8a66d3f00 dir:0
[W][000075437.016108][audioconvert.c:1008 impl_node_port_use_buffers()] 0x7f799b516038: USE BUFFERS 0x55d8a66d40b0 dir:0
[W][000075437.016580][audioconvert.c:957 impl_node_port_set_param()] 0x7f799b516038: SET PARAM:4 DIR:1 param:0x55d8a6596950
[W][000075437.016593][audioconvert.c:971 impl_node_port_set_param()] 0x7f799b516038:     clean 1, res 0
[W][000075437.017421][audioconvert.c:957 impl_node_port_set_param()] 0x7f79a51ce228: SET PARAM:4 DIR:0 param:0x55d8a6596950
[W][000075437.017435][audioconvert.c:971 impl_node_port_set_param()] 0x7f79a51ce228:     clean 0, res 0
[W][000075437.583973][audioconvert.c:957 impl_node_port_set_param()] 0x7f79a51ce228: SET PARAM:4 DIR:1 param:0x7ffc8237e618
[W][000075437.584002][audioconvert.c:971 impl_node_port_set_param()] 0x7f79a51ce228:     clean 1, res 0
[W][000075437.607574][audioconvert.c:1008 impl_node_port_use_buffers()] 0x7f799b516038: USE BUFFERS 0x55d8a66e6350 dir:1
[W][000075437.607591][audioconvert.c:1008 impl_node_port_use_buffers()] 0x7f79a51ce228: USE BUFFERS 0x55d8a66e6350 dir:0
[W][000075437.607616][audioconvert.c:1008 impl_node_port_use_buffers()] 0x7f79a51ce228: USE BUFFERS 0x55d8a66f6e60 dir:1
[W][000075437.607687][audioconvert.c:1008 impl_node_port_use_buffers()] 0x7f79a51ce228: USE BUFFERS 0x7f79a1fa0080 dir:0
```

`0x7f799b516038` is the audioconvert node
`0x7f79a51ce228` is the adapter

So, basically, buffers are configured on the audioconvert  INPUT ports right after the format is set, but then the format is set for the OUTPUT port, which is in "convert" mode, and the `buffers_set[INPUT]` flag is cleared, so when the buffers are configured for the OUTPUT port, `setup_buffers` is not triggered.

**Comment by gkiagia** [comment]
I opened this ticket instead of fixing the code directly, because I am not sure what the intention of the code is.

One easy solution would be to change:
```
if (id == SPA_PARAM_Format && !is_monitor) {
		if (param == NULL ||
		    this->mode[direction] == SPA_PARAM_PORT_CONFIG_MODE_convert)
			clean_convert(this);
	}
```
to
```
if (id == SPA_PARAM_Format && !is_monitor) {
		if (param == NULL)
			clean_convert(this);
	}
```

Another solution would be to not clear `buffers_set` inside `clean_convert` and clear it somewhere else (not sure where)

**Comment by wtay** [comment]
I think I fixed it with the above patch
